tests/xharness/AppRunner.cs

```diff
@@ -572,57 +352,28 @@ void CleanXml (string source, string destination)
 			// from the TCP connection, we are going to fail when trying to read it and not parse it. Therefore, we are not only
 			// going to check if we are in CI, but also if the listener_log is valid.
 			var path = Path.ChangeExtension (test_log_path, "xml");
-			CleanXml (test_log_path, path);
+			XmlResultParser.CleanXml (test_log_path, path);
+			XmlResultParser.Jargon xmlType;
 
-			if (Harness.InCI && IsXml (test_log_path)) {
+			if (Harness.InCI && XmlResultParser.IsValidXml (path, out xmlType)) {
 				(string resultLine, bool failed, bool crashed) parseResult = (null, false, false);
 				crashed = false;
-				XmlResultType xmlType;
 				try {
-					using (var streamReaderTmp = new StreamReader (path)) {
-						xmlType = GetXmlType (streamReaderTmp);
-						var fileName = Path.GetFileName (path);
-						var newFilename = "";
-						switch (xmlType) {
-						case XmlResultType.TouchUnit:
-						case XmlResultType.NUnit:
-							newFilename = path.Replace (fileName, $"nunit-{fileName}");
-							break;
-						case XmlResultType.xUnit:
-							newFilename = path.Replace (fileName, $"xunit-{fileName}");
-							break;
-						}
-						// rename the path to the correct value
-						File.Move (path, newFilename);
-						path = newFilename;
-						
-					}
-					using (var reader = new StreamReader (path)) {
-						var tmpFile = Path.Combine (Path.GetTempPath (), Guid.NewGuid ().ToString ());
-						using (var writer = new StreamWriter (tmpFile, true)) { // write the human result to the log file
-							(string resultLine, bool failed) parseData;
-							switch (xmlType) {
-							case XmlResultType.TouchUnit:
-								parseData = ParseTouchUnitXml (reader, writer);
-								break;
-							case XmlResultType.NUnit:
-								parseData = ParseNUnitXml (reader, writer);
-								break;
-							case XmlResultType.xUnit:
-								parseData = ParsexUnitXml (reader, writer);
-								break;
-							default:
-								parseData = ("", true);
-								break;
-							}
-							parseResult.resultLine = parseData.resultLine;
-							parseResult.failed = parseData.failed;
-						}
-						File.Copy (tmpFile, test_log_path, true);
-					}
+					var newFilename = XmlResultParser.GetXmlFilePath (path, xmlType);
+
+					// rename the path to the correct value
+					File.Move (path, newFilename);
+					path = newFilename;
+
+					// write the human readable results in a tmp file, which we later use to step on the logs
```

Comment:
Could we instead add both files to the html report? Replacing/rewriting files ends up being confusing.

Reply:
Both are added. Take a look at: https://github.com/xamarin/xamarin-macios/pull/7794/files/46600f2ba5ddc491bc58530e93b4d4509f00af04#diff-03773861bcef485bf343300f31b60b0eR374

We step on the xml written by the log and add it as the human readable version, later we add a new xml log. The final html report looks like this:

<img width="934" alt="Screenshot 2020-02-04 at 14 11 07" src="https://user-images.githubusercontent.com/2190086/73777882-3a5ba480-4758-11ea-842c-c3730e8e49e6.png">

Clicking on 'Test Log' shows the human readable ones. Xml shows the xml (useful for debugging or importing somewhere).

Reply:
This still seems more complex than needed to me. Why not just write out the human readable version to a new file and add that, without stepping on anything we already have?

Also, when is `tmpFile` deleted? The usual way to create a log file is to use `Logs.CreateFile`, which will create a file next to the other logs, and which will also be deleted whenever those are deleted (during a `git clean` for instance).

Reply:
The TestLog is badly formatted xml (include the ping and other stuff). We clean it and create the xml file. At that point the Test Log is no very useful for a human, so we step on it with the human readable version. 

It is true that I forgot to delete the tmp file.

Reply:
> At that point the Test Log is no very useful for a human

It's useful for unconfusing humans who wonder why they don't get the same results / behavior when running tests locally vs running tests on CI :smile:

Reply:
@rolfbjarne but this just happens on CI where we have xml, not locally. We could set to have xml ALWAYS, which is another option to be the same

Reply:
> but this just happens on CI where we have xml, not locally.

Yes, that's fine. The confusing part is that replacing files with different content ends up being confusing, especially when something goes wrong and only some files are replaced and then you have several files that are named similarly but with wildly different content.

Keeping all the files around avoids this.

Reply:
Ok, I now understand, I created https://github.com/xamarin/xamarin-macios/issues/7802 lets keep it there, get it to work and I'll do the clean up later. I want to get the CI ready ASAP.